include/irr/asset/CTexturePacker.h

```diff
@@ -105,31 +84,74 @@ class ITexturePacker
     {
-        auto extent = _img->getCreationParameters().extent;
-
-        const uint32_t w = std::max(extent.width>>_subres.baseMipLevel,1u);
-        const uint32_t h = std::max(extent.height>>_subres.baseMipLevel,1u);
-        const uint32_t pgSqrSz = core::roundUpToPoT((std::max(w,h)+m_pgSzxy-1u)/m_pgSzxy);
+        const uint32_t pgSqrSz = computeSquareSz(_img, _subres);
 
         const uint32_t addr = m_pgtAddrAlctr.alloc_addr(pgSqrSz);
         return (addr==CSquareAddressAllocator<uint32_t>::invalid_address) ? 
             page_tab_offset_invalid() :
             page_tab_offset_t(m_pgtAddrAlctr.unpackAddress_x(addr), m_pgtAddrAlctr.unpackAddress_y(addr));
     }
+    virtual void free(page_tab_offset_t _addr, const ICPUImage* _img, const ICPUImage::SSubresourceRange& _subres)
+    {
+        if ((_addr==page_tab_offset_invalid()).all())
+            return;
+
+        const uint32_t pgSqrSz = computeSquareSz(_img, _subres);
+
+        m_pgtAddrAlctr.free_addr(core::morton2d_encode(_addr.x,_addr.y), pgSqrSz);
+    }
+
+private:
+    uint32_t computeSquareSz(const ICPUImage* _img, const ICPUImage::SSubresourceRange& _subres)
+    {
+        auto extent = _img->getCreationParameters().extent;
+
+        const uint32_t w = std::max(extent.width>>_subres.baseMipLevel,1u);
+        const uint32_t h = std::max(extent.height>>_subres.baseMipLevel,1u);
+
+        return core::roundUpToPoT((std::max(w,h)+m_pgSzxy-1u)/m_pgSzxy);
+    }
 };
 
 class ICPUTexturePacker : protected ITexturePacker
 {
-    uint32_t physAddrTexLayerSz(uint32_t _pgSz, uint32_t _tilesPerDim) const
+public:
+    //! SPhysPgOffset is what is stored in texels of page table!
+    struct SPhysPgOffset
+    {
+        _IRR_STATIC_INLINE_CONSTEXPR uint32_t invalid_addr = 0xffffu;
+
+        SPhysPgOffset(uint32_t _addr) : addr(_addr) {}
+
+        //upper 16 bits are used for address of mip-tail page
+        uint32_t addr;
+    };
+    SPhysPgOffset makeTexOffset(uint32_t _x, uint32_t _y, uint32_t _layer)
+    {
+        uint32_t addr = (_x & PHYS_PG_OFFSET_X_MASK);
+        addr |= (_y & PHYS_PG_OFFSET_X_MASK) << (ADDR_LAYER_SHIFT >> 1);
+        addr |= (_layer << ADDR_LAYER_SHIFT);
+        addr |= (SPhysPgOffset::invalid_addr << 16);
+        return addr;
+    }
+    uint32_t physPgOffset_x(SPhysPgOffset _offset) const { return _offset.addr & PHYS_PG_OFFSET_X_MASK; }
+    uint32_t physPgOffset_y(SPhysPgOffset _offset) const { return (_offset.addr >> (ADDR_LAYER_SHIFT>>1)) & PHYS_PG_OFFSET_X_MASK; }
+    uint32_t physPgOffset_layer(SPhysPgOffset _offset) const { return (_offset.addr & 0xffffu)>>ADDR_LAYER_SHIFT; }
+    bool physPgOffset_valid(SPhysPgOffset _offset) const { return (_offset.addr&0xffffu) != SPhysPgOffset::invalid_addr; }
+    bool physPgOffset_hasMipTailAddr(SPhysPgOffset _offset) const { return physPgOffset_valid(physPgOffset_mipTailAddr(_offset)); }
+    SPhysPgOffset physPgOffset_mipTailAddr(SPhysPgOffset _offset) const { return _offset.addr>>16; }
```

Comment:
you can share all this between `ICPUTexturePacker` and a future `IGPUTexturePacker` by declaring it all in `ITexturePacker`